File: 2870-minimum-number-of-operations-to-make-array-empty/2870-minimum-number-of-operations-to-make-array-empty.swift
class Solution {
    func minOperations(_ nums: [Int]) -> Int {
        var result = 0
        
        var occ = [Int: Int]()
        for num in nums {
            guard occ[num] != nil else {
                occ[num] = 1
                continue
            }
            
            occ[num]! += 1
        }
        
        print(occ)

        for occurence in occ.values {
            let partial = breakDown(occurence)
            
            if partial > 0 {
                result += partial
            } else {
                return -1
            }
        }
        
        return result
    }
    
    func breakDown(_ number: Int) -> Int {        
        if number % 3 == 0 {
            return number / 3
        } 
        
        else if number % 3 == 2 {
            return (number / 3) + 1
        }
        
        else if number % 3 == 1 && number / 3 > 1 {
            return (number - 4) / 3 + 2
        }
        
        else if number % 2 == 0 {
            return number / 2
        } 
        
        else {
            return -1
        }
    }
}

File: 2870-minimum-number-of-operations-to-make-array-empty/2870-minimum-number-of-operations-to-make-array-empty.py
from collections import defaultdict

class Solution:
    def breakdown (self, n: int) -> int:
        if n % 3 == 0:
            return n // 3
        elif n % 3 == 2:
            return (n // 3) + 1
        elif n % 3 == 1 and n // 3 > 1:
            return ((n - 4) // 3) + 2
        elif n % 2 == 0:
            return n // 2
        else:
            return -1
    
    def minOperations(self, nums: List[int]) -> int:
        occ = defaultdict(int)
        
        for num in nums:
            occ[num] += 1
        
        result = 0
        for value in occ.values():
            breaked = self.breakdown(value)
            if breaked < 0:
                return -1
            else:
                result += breaked
        
        return result